src/catch2/catch_matchers_templates.hpp

```diff
@@ -0,0 +1,257 @@
+        template<typename... MatcherTs> struct MatchAnyOfGeneric;
+        template<typename MatcherT> struct MatchNotOfGeneric;
+
+        struct MatcherGenericBase : MatcherUntypedBase {};
+
+        std::vector<void const*> vector_cat(std::vector<void const*> const & lhs, std::vector<void const*> const& rhs);
+        std::vector<void const*> vector_cat(std::vector<void const*> const & lhs, void const* rhs);
+        std::vector<void const*> vector_cat(void const* lhs, std::vector<void const*> const& rhs);
+
+
+        #ifdef CATCH_CPP17_OR_GREATER
+
+        using std::conjunction;
+
+        #else // CATCH_CPP17_OR_GREATER
+
+        template<typename... Cond>
+        struct conjunction : std::true_type {};
+
+        template<typename Cond, typename... Rest>
+        struct conjunction<Cond, Rest...> : std::integral_constant<bool, Cond::value && conjunction<Rest...>::value> {};
+
+        #endif // CATCH_CPP17_OR_GREATER
+
+
+        using std::index_sequence;
+        using std::index_sequence_for;
+        using std::make_index_sequence;
+
+
+        template<typename T>
+        using is_generic_matcher = std::is_base_of<
+            Catch::Matchers::Impl::MatcherGenericBase,
+            typename std::remove_cv<typename std::remove_reference<T>::type>::type
+        >;
+
+        template<typename... Ts>
+        using are_generic_matchers = conjunction<is_generic_matcher<Ts>...>;
+
+        template<typename T>
+        using is_matcher = std::is_base_of<
+            Catch::Matchers::Impl::MatcherUntypedBase,
+            typename std::remove_cv<typename std::remove_reference<T>::type>::type
+        >;
+
+
+        template<typename..., typename Arg>
+        bool match_all_of(Arg const&, std::vector<void const*> const&, index_sequence<>) {
+            return true;
+        }
+
+        template<typename T, typename... MatcherTs, typename Arg, std::size_t Idx, std::size_t... Indices>
+        bool match_all_of(Arg const& arg, std::vector<void const*> const& matchers, index_sequence<Idx, Indices...>) {
+            return static_cast<T const*>(matchers[Idx])->match(arg) && match_all_of<MatcherTs...>(arg, matchers, index_sequence<Indices...>{});
+        }
+
+
+        template<typename..., typename Arg>
+        bool match_any_of(Arg const&, std::vector<void const*> const&, index_sequence<>) {
+            return false;
+        }
+
+        template<typename T, typename... MatcherTs, typename Arg, std::size_t Idx, std::size_t... Indices>
+        bool match_any_of(Arg const& arg, std::vector<void const*> const& matchers, index_sequence<Idx, Indices...>) {
+            return static_cast<T const*>(matchers[Idx])->match(arg) || match_any_of<MatcherTs...>(arg, matchers, index_sequence<Indices...>{});
+        }
+
+
+        template<typename... MatcherTs, std::size_t... Idx>
+        std::string describe_multi_matcher(std::string const& combine, std::vector<void const*> const& matchers, index_sequence<Idx...>) {
```

Comment:
I would also consider using `StringRef` together with its UDL for the `combine` arguments, as we then don't have to rely on good implementation of `std::string` to avoid allocations.

Reply:
will do